lib/Twig/Extension/Escaper.php

```diff
@@ -87,3 +206,281 @@ function twig_raw_filter($string)
 {
     return $string;
 }
+
+/**
+ * Escapes a string.
+ *
+ * @param Twig_Environment $env
+ * @param mixed            $string     The value to be escaped
+ * @param string           $strategy   The escaping strategy
+ * @param string           $charset    The charset
+ * @param bool             $autoescape Whether the function is called by the auto-escaping feature (true) or by the developer (false)
+ *
+ * @return string
+ */
+function twig_escape_filter(Twig_Environment $env, $string, $strategy = 'html', $charset = null, $autoescape = false)
+{
+    if ($autoescape && $string instanceof Twig_Markup) {
+        return $string;
+    }
+
+    if (!is_string($string)) {
+        if (is_object($string) && method_exists($string, '__toString')) {
+            $string = (string) $string;
+        } elseif (in_array($strategy, array('html', 'js', 'css', 'html_attr', 'url'))) {
+            return $string;
+        }
+    }
+
+    if (null === $charset) {
+        $charset = $env->getCharset();
+    }
+
+    $escapers = $env->getExtension('Twig_Extension_Escaper')->getEscapers();
```

Comment:
Current code is optimized for speed. It looks like the new one is not. That's a very sensitive part of the code as it might be called thousand of times for a page, and small changes in the past had great performance impact.